tools/src/bin/scip-indexer.rs

```diff
@@ -0,0 +1,1222 @@
+    let byte_count = file.metadata().expect("Failed to get file metadata").len();
+    let mut file = BufReader::new(file);
+    let mut file = protobuf::CodedInputStream::from_buf_read(&mut file);
+    let index = Index::parse_from(&mut file).expect("Failed to read scip index");
+
+    let mut scip_symbol_to_structured: HashMap<String, AnalysisStructured> = HashMap::new();
+    let mut our_symbol_to_scip_sym: UstrMap<String> = UstrMap::default();
+
+    let (lang_name, lang) = match index.metadata.tool_info.name.as_str() {
+        "rust-analyzer" => ("rs", ScipLang::Rust),
+        "scip-python" => ("py", ScipLang::Python),
+        "scip-typescript" => ("js", ScipLang::Typescript),
+        _ => {
+            warn!("Unsupported language; we need tree-sitter support.");
+            return;
+        }
+    };
+
+    // ## First Pass: Process Symbol Definitions
+    //
+    // It's necessary to process all of the symbol definitions first because
+    // occurrences can reference symbols defined in other documents.  (The
+    // exception is locals inherently are file-local.)
+    //
+    // This also provides us a good opportunity to normalize the symbol names we
+    // see and perform structural inference from the descriptors.  Structural
+    // inference inherently involves mutating symbols as we go, so we do not
+    // write anything out during this pass.  This is just as well as we can make
+    // sure to only emit structured information at the point of definition
+    // (which we do not know until we look at the occurrences).
+    for doc in &index.documents {
+        info!(
+            "Processing symbols/definitions for '{}'",
+            &doc.relative_path,
+        );
+
+        // XXX next steps
+        // finish up the loop before to populate the structured things:
+        // - was thinking we just re-derive the descriptor string ourselves as
+        //   we go and we hold onto the previous state of the derived string so
+        //   that we can do the parent lookup for fields/methods.
+        // - the previous descriptor isn't particularly useful because it's
+        //   always going to be a type.
+        // - extra things:
+        //   - we do want to handle the relationship
+        //   - it could be good to have a basic approach for the documentation
+        //     - for rust that does give us size and alignment which is nice
+        //     - for JS we get the inferred type as well, plus the extracted
+        //       comment as a second string after the type info.
+
+        for scip_sym_info in &doc.symbols {
+            // Process each symbol to:
+            // - Derive a canonical mozsearch symbol name and map it.
+            // - Derive structured analysis information that we will emit in the
+            //   next pass if/when we see the definition.
+            if let Ok(scip_sym) = scip::symbol::parse_symbol(&scip_sym_info.symbol) {
```

Comment:
What do you think of factoring this out / deindent this a little bit?

Something like:

```
let Ok(scip_sym) = scip::symbol::parse_symbol(..) else {
    warn!(...);
    continue;
};

// ### Extract M...
```

Reply:
Oh wow, I'm not sure I realized [let-else](https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html) existed!  This is fantastic!  I'll definitely adapt this here and potentially there are some other spots where I currently use a match-based idiom for the same purpose that could be made more concise with let-else,

Thanks!

Reply:
Amending this, to avoid causing bit-rot for @nicolas-guichard I'm not going to do this right now, but will be very happy if @nicolas-guichard wants to make changes like this.